### ISSUE TITLE
[discussion] Native unicode support

### ISSUE THREAD

**Issue by avih** [rmyorston/busybox-w32#352]
So, I'm looking into adding native unicode support.

Basically similar to what the utf8 manifest does, but manually (adding `U` variant of win32 APIs which use utf8 prototypes and then internally calls the `W` APIs - instead of letting the manifest change the `A` APIs into utf8).

So I created this issue to ask related questions instead of opening an issue on each question.

First question: I might need to maintain a utf8 version of `environ` independent of the system `environ` (with its own *env API). Does busybox-w32 expect `putenv` to add the string to the environment, so that e.g. this prints `1` ?

```c
char xenv[] = "x=0";
putenv(xenv);
xenv[2] = '1';
printf("%s\n", getenv("x"));
```

With POSIX *env API (and on linux) it prints 1, but on windows, as far as I can tell `_putenv` makes a copy of `xenv`, so the user's string doesn't actually become part of the environment, and it prints 0.

Also, with POSIX, it's valid to change `environ` to point to something else (user-provided array, for instance).

I see that `mingw_putenv` does a trick to set an empty value, and it's indeed aware that the buffer gets copied (hence `environ` is iterated to find the copy to truncate).

Do you know if these use cases (change the buffer after `putenv`, or change `environ`) happen with upstream busybox? If yes, how does it work on windows?

(`setenv` does make a copy, and so it can be more efficient to access using something less linear than `environ`).

**Comment by rmyorston** [comment]
I'm not _aware_ of any case which relies on being able to change the buffer after `putenv`.

`ash` changes `environ` in `evalcommand()` when running a nofork applet. The existing value of `environ` is saved and restored after the applet completes.

**Comment by avih** [comment]
Thanks.

I ended up (ab)using the ANSI `environ` to store the UTF8 values, same as I did 3 years ago here https://github.com/rmyorston/busybox-w32/issues/17#issuecomment-681927971 so basically the native unicode thing uses the ANSI `environ` normally as far as the system goes, just that the encoding of non-ascii values is utf8 instead of ACP (and the mingw getenv/putenv/etc wrappers remain as is).

So any trick busybox or busybox-w32 do (like with the empty values, or the clearenv implementation), _should_ work out of the box also with the native unicode mode.

> `ash` changes `environ` in `evalcommand()`

As far as I can tell, when spawn is used - it takes the process env - not the CRT `environ`, but the process env is hopefully up to date because `putenv` updates both the crt `environ` and also the process env (e.g. `SetEnvironmentVariable[W]`).

But if you modify `environ` to point to something else, how does the process env gets updated to the new values?

Still on the subject of `environ`.

I see that there were some commits related to UCRT. As far as I can tell, the current state is that we can't spawn with UCRT with non-null `env`, or else it will crash?

And so, if we need to spawn with a custom `env` array, it basically clears the current env and sets it from this custom array, and then calls spawn with a `NULL` env so that it would be taken from the environment (which is now a copy of the custom array).

Is that correct?

Originally I meant to handle NULL env to spawn by simply replacing it with `environ` (which is then converted to a wide array before calling `_wspawnve`), so that would presumably crash with UCRT?

Can you reproduce the crash issue with NULL env? Any pointers how to build busybox-w32 with UCRT?

**Comment by rmyorston** [comment]
Nofork applets run in the same process as their parent shell. There's no spawn in this case.

Passing a non-NULL env to spawn with UCRT causes any subsequent spawn in the child process to fail.  If the child process _doesn't_ spawn a grandchild there's no problem. The workaround is to arrange things so that the initial spawn always has a NULL env.

I haven't reproduced this recently. When I worked on the problem I used the UCRT variant of MSYS2 on Windows.

**Comment by avih** [comment]
> Passing a non-NULL env to spawn with UCRT causes any subsequent spawn in the child process to fail. If the child process _doesn't_ spawn a grandchild there's no problem. The workaround is to arrange things so that the initial spawn always has a NULL env.

Thanks.

So how does one trigger the crash scenario as far as a user test case goes?

Also, I might want to put all the UTF8 APIs implementations in a new C file (maybe `mingw-utf8.c` or some such), possibly with a matching header.

How should I go about it as far as makefiles/other-requirements go? I tried grepping `mingw.c` and `mingw.o` (to check how `mingw.c` is compiled depending on CONFIG_PLATFORM_MINGW32) but couldn't find how this happens.

**Comment by rmyorston** [comment]
>So how does one trigger the crash scenario as far as a user test case goes?

You'd need to:
- Revert the fix in the shell: `sed -i 's/_UCRT/_WHATEVER/' shell/ash.c`
- Build busybox-w32 using the UCRT variant of MSYS2
- Have the shell spawn an application that spawns some other application. In the original issue the first application was GNU make and I used busybox-w32 `time` as the second.

>Do I need to mention the .h file someplace

That shouldn't be necessary.

**Comment by avih** [comment]
Thanks.

I still didn't get to try UCRT, but I'll update here once I do.

Regardless, before you make the next release, can I please review the release notes about unicode?

**Comment by rmyorston** [comment]
[Draft release notes](https://frippery.org/busybox/release-notes/FRP-5179.html)

**Comment by avih** [comment]
> Draft release notes

Thanks. Generally looks good. Few minor comments:

> It isn't necessary to adjust the console code page: this is handled transparently.

That might sound as if we're changing the console CP. Assuming we don't want to imply this, maybe remove "this is handled transparently", or maybe reword as something like "Unicode output and interactive input work with any console code page."

> It has been necessary to apply workarounds for deficiencies in Microsoft's current support for UTF-8. There may be undiscovered problems and the workarounds themselves may have issues.

I believe this refers only to UTF8_INPUT and/or UTF8_OUTPUT? (I don't recall other workarounds)

If yes, then it might be useful to use something like "... deficiencies in Microsoft's current support for UTF-8 in interactive use and printouts to the console"

Or some other wording to limit the statement to console IO, as this doesn't refer, as far as I can tell, to the ability to work with files with unicode names, like with `tar` or `cat`.

> The Windows Terminal works better than the Windows console.

Maybe add "especially when it comes to rendering unicode". I don't _think_ there's difference in editing as long as the glyphs are supported.

As an additional note, I found out that other terminals also work better than the windows console, like vscode builtin terminal (amazing combining chars rendering), or [hyper](https://github.com/vercel/hyper) or [wez](https://github.com/wez/wezterm), and possibly more.

All the mentioned ones seem to handle glyph widths better than the windows terminal, tested with [this](https://www.w3.org/2001/06/utf-8-test/UTF-8-demo.html) - check "The Three Kingdoms" alignment thingy.

But I agree that mentioning only the windows terminal is enough.

> Editing (cursor movement, delete, etc) can be buggy, especially with double-width or combining characters.

A reminder that while it doesn't fix all busybox editing issues, I do have a version with `wcwidth` updated to latest Unicode (15). We already pushed a tiny update only for emojis, but the full update has much more. I can arrange quickly a PR with the full update if you want, which will be [this](https://github.com/avih/busybox-w32/commit/4755f6d9445cd15005a78f6b6d6eddf58178caf0), but it will use it unconditionally with UTF8_MANIFEST (and disable the existing table). Or just feel free to take this commit, edit it as you see fit, and push it yourself.

I don't know how much it affects practical use cases, but I believe it would affect that rocketship prompt which was mentioned in another issue, because that rocket glyph is double width, but current busybox[-w32] doesn't know that, so if it appears as part of PS1 at the last line of the prompt, busybox will get confused with the line wrap (of the user input which follows). The full update to Unicode 15 does include it as double width.

> `(mingw64-gcc 12.2.1-8.fc38; mingw64-crt 10.0.0-4.fc38; glob; Unicode)`

As I think it's feasible to get the native unicode working, maybe change it to `Unicode-manifest` or `Unicode-Win10` so that the difference becomes visible (I'm not suggesting again to also change the download file name).

Regardless of Unicode, this looks like another good release, and good notes. Thanks again for maintaining it.

**Comment by rmyorston** [comment]
I originally wrote
>It isn't necessary to adjust the console code page: this is handled automatically.

but thought that might imply the code page was changed automatically. So I said 'transparently' instead.

End users probably don't care too much about the exact nature of Microsoft's deficiencies so I prefer to leave them unspecified.

I'll have a look at the `wcwidth` commit later.

The release certainly isn't going to be FRP-5179. Testing has uncovered problems on Wine which will require at least one workaround.

**Comment by avih** [comment]
> I'll have a look at the `wcwidth` commit later.

Sure. Let me know if you need anything.

> The release certainly isn't going to be FRP-5179. Testing has uncovered problems on Wine which will require at least one workaround.

Hmm... I presume that's in your provate test setup?

> However, all of those also don't have mouse-wheel events for console apps which support it (like `less` for windows), while such apps do work at the windows terminal/console.

FWIW, this seems to be an issue of older (win10, but not win11) `conpty.dll`, which apparently doesn't do mouse events well.

But alacritty master supports loading a newer `conprt.dll`.

As of now, one would need a 3rd party alacritty nightly build, like from [here](https://github.com/alacritty/alacritty/pull/6994#issuecomment-1619432716), and then some 3rd party build of `conpty.dll` and `OpenConsole.exe` like [here](https://github.com/wez/wezterm/tree/main/assets/windows/conhost) because these files apparently are not packaged with latest windows terminal.

So if they're at the same dir, alacritty.exe will load it, and then console applications get mouse events, seemingly like they do in windows console/terminal.

**Comment by rmyorston** [comment]
I've decided not to take the `wcwidth` commit for now. I always seem to have cause to regret last-minute changes made just before a release.

I'm not impressed with the latest version of Wine in Fedora.  There are a number of regressions. I've worked around the most serious; the rest can wait. It's not like Wine is an important platform for busybox-w32.

>Hmm... I presume that's in your private test setup?

Not private: `testsuite/runtest` in busybox-w32 works on Windows as well as Linux. There are more failures on Windows, but that's only to be expected.

I'll restart my testing with FRP-5181 and see how that goes.

**Comment by avih** [comment]
> I've decided not to take the `wcwidth` commit for now

Sure. But if the unicode build gains traction, then I think it would be good to get this or something similar in. The upstream `wcwidth` is 15 years out of date, and the number of meme double-width codepoints/icons has grown considerably since then - both in official support and in various meme use cases, like very fancy PS1.

I'm not suggesting to push it upstream because TBH I don't think we should deal with the "compression" of the data at upstream wcwidth whenever a new Unicode version is released (I think Unicode 16 should be released in few weeks).

I don't think there are existing scripts to automate it from the Unicode data to the busybox "compression", so taking it from someplace which does automate it sounds a lot better to me, especially where we might have slightly less concerns with the balance of program size vs features.

> Not private: `testsuite/runtest` in busybox-w32 works on Windows as well as Linux. There are more failures on Windows, but that's only to be expected.

Do the linux parts refer to testing busybox-w32 on linux using wine? or to testing the native linux busybox?

Anyway, back to the subject of this issue: native unicode support.

I've noticed that in some win32 files, `libbb.h` is included before other standard headers, for instance at `fsync.c` or `glob.c` or `poll.c` or even `mingw.c`.

Now, `libbb.h` includes `mingw.h`, and `mingw.h` has some mappings of default names, such as `fopen` to `mingw_fopen`, etc.

So the point is that I think `libbb.h` should be included last, so that it doesn't overrides system names before their respective system header is included.

I don't think I can identify a reason it's included first. Is there such reason?

If this is expected to work because `libbb.h` already includes these headers, and so they'll no-op the second `#include`, then I think it's kind of meh IMO, but would still be good to know.

I'm asking this because the utf8 API would need similar global mapping which would probably apply via `mingw.h` (like `CreateFileA` to `createFile_U`), but if `windows.h` is included after `libbb.h`, then this poses an issue (and same for other names which would require such mappings).

(of course, names which are already mapped globally currently, like `mingw_fopen`, would not have global mapping to the utf8 variants. These name would map to the utf8 variants at the C file where the current mingw/winansi wrapper is implemented).

**Comment by rmyorston** [comment]
>Do the linux parts refer to testing busybox-w32 on linux using wine? or to testing the native linux busybox?

The latter. It's possible to build and test a native Linux binary using the busybox-w32 source. The test output doesn't exactly match that obtained using the upstream source. That's because some Windows tests are included when they really shouldn't be. I'll fix that.

`libbb.h` tends to be included first because that's how it's always been done:
```
fa37e9da4b (Nguyễn Thái Ngọc Duy 2010-04-14 02:12:01 +0200    5) #include "libbb.h"
fa37e9da4b (Nguyễn Thái Ngọc Duy 2010-04-14 02:12:01 +0200    6) #include <windows.h>
1ee308c75f (Ron Yorston          2021-12-22 08:22:12 +0000    7) #include "lazyload.h"
fa37e9da4b (Nguyễn Thái Ngọc Duy 2010-04-14 02:12:01 +0200    8) #undef PACKED
fa37e9da4b (Nguyễn Thái Ngọc Duy 2010-04-14 02:12:01 +0200    9)
```
I've just been following that existing practice.

**Comment by avih** [comment]
> I've just been following that existing practice.

Would you be against moving `libbb.h` it to the end?

For instance, mapping `fprintf` to `mingw_printf` would be disastarous if this happens before `fprintf` is declared...

Anyway, next subject.

I was trying to evaluate which/how-many APIs would need to be mapped from ansi to UTF8. Searching APIs at the code did not seem productive, so I wrote a script which produces this list of the ansi symbols used by `busybox.exe`.

<details>
<summary>List of ansi symbols used by `busybox.exe` (about 100)</summary>

```
$ w32sym.sh -A busybox.exe
GetUserNameA
CreateEventA
CreateFileA
CreateFileMappingA
CreateNamedPipeA
CreateProcessA
FillConsoleOutputCharacterA
FindFirstFileA
FindNextFileA
GetCompressedFileSizeA
GetDiskFreeSpaceExA
GetDriveTypeA
GetFileAttributesA
GetFileAttributesExA
GetFullPathNameA
GetModuleFileNameA
GetModuleHandleA
GetSystemDirectoryA
GetVersionExA
GetVolumeInformationA
LoadLibraryA
LoadLibraryExA
MoveFileExA
Process32First
Process32Next
ReadConsoleInputA
SetConsoleTitleA
SetEnvironmentVariableA
SetFileAttributesA
__argv
__getmainargs
__initenv
_access
_ctime64
_environ
_fdopen
_fullpath
_getch
_open
_putenv
_vscprintf
_vsnprintf
fgets
fopen
fprintf
fputc
fputs
getenv
isalnum
isalpha
iscntrl
isgraph
islower
isprint
ispunct
isspace
isupper
isxdigit
putc
putchar
rename
setlocale
sprintf
sscanf
strcat
strchr
strcmp
strcpy
strcspn
strftime
strlen
strncmp
strncpy
strpbrk
strrchr
strspn
strstr
strtok
strtol
strtoul
tolower
toupper
ungetc
vfprintf
_unlink
_spawnve
_rmdir
_open
_mktemp
_mkdir
_getcwd
_getche
_fdopen
_creat
_chmod
_chdir
DispatchMessageA
PeekMessageA
WSASocketA
```
</details>

<details>
<summary>List of all symbols used by `bysybox.exe`, with the ansi ones marked, with the respective wide symbol</summary>

```
$ w32sym.sh -a busybox.exe
ADVAPI32.dll:
    AccessCheck
    CheckTokenMembership
    DuplicateToken
    EqualSid
    GetSecurityInfo
    GetTokenInformation
[A] GetUserNameA (GetUserNameW)
    OpenProcessToken
    SaferComputeTokenFromLevel
    SaferCreateLevel
    SetTokenInformation
    SystemFunction036

KERNEL32.dll:
    CloseHandle
    CreateConsoleScreenBuffer
[A] CreateEventA (CreateEventW)
[A] CreateFileA (CreateFileW)
[A] CreateFileMappingA (CreateFileMappingW)
[A] CreateNamedPipeA (CreateNamedPipeW)
    CreatePipe
[A] CreateProcessA (CreateProcessW)
    CreateRemoteThread
    CreateToolhelp32Snapshot
    DeleteCriticalSection
    DeviceIoControl
    DuplicateHandle
    EnterCriticalSection
    FileTimeToSystemTime
    FillConsoleOutputAttribute
[A] FillConsoleOutputCharacterA (FillConsoleOutputCharacterW)
    FindClose
[A] FindFirstFileA (FindFirstFileW)
    FindFirstVolumeW
[A] FindNextFileA (FindNextFileW)
    FindNextVolumeW
    FindVolumeClose
    FlushFileBuffers
    FreeEnvironmentStringsW
    GenerateConsoleCtrlEvent
    GetACP
    GetCPInfo
    GetCommandLineW
[A] GetCompressedFileSizeA (GetCompressedFileSizeW)
    GetConsoleCP
    GetConsoleMode
    GetConsoleOutputCP
    GetConsoleScreenBufferInfo
    GetConsoleWindow
    GetCurrentProcess
[A] GetDiskFreeSpaceExA (GetDiskFreeSpaceExW)
    GetDiskFreeSpaceExW
[A] GetDriveTypeA (GetDriveTypeW)
    GetEnvironmentStringsW
    GetEnvironmentVariableW
    GetExitCodeProcess
[A] GetFileAttributesA (GetFileAttributesW)
[A] GetFileAttributesExA (GetFileAttributesExW)
    GetFileInformationByHandle
    GetFileSizeEx
    GetFileType
[A] GetFullPathNameA (GetFullPathNameW)
    GetLastError
    GetLogicalDrives
[A] GetModuleFileNameA (GetModuleFileNameW)
[A] GetModuleHandleA (GetModuleHandleW)
    GetNumberOfConsoleInputEvents
    GetProcAddress
    GetProcessAffinityMask
    GetProcessId
    GetProcessTimes
    GetStdHandle
[A] GetSystemDirectoryA (GetSystemDirectoryW)
    GetSystemInfo
    GetSystemTimeAsFileTime
    GetTickCount
[A] GetVersionExA (GetVersionExW)
[A] GetVolumeInformationA (GetVolumeInformationW)
    GetVolumeInformationW
    InitializeCriticalSection
    IsDBCSLeadByteEx
    IsValidCodePage
    IsWow64Process
    LeaveCriticalSection
[A] LoadLibraryA (LoadLibraryW)
[A] LoadLibraryExA (LoadLibraryExW)
    LocalFree
    MapViewOfFile
[A] MoveFileExA (MoveFileExW)
    MultiByteToWideChar
    OpenProcess
    PeekConsoleInputW
    PeekNamedPipe
[A] Process32First (Process32FirstW)
[A] Process32Next (Process32NextW)
[A] ReadConsoleInputA (ReadConsoleInputW)
    ReadConsoleInputW
    ReadDirectoryChangesW
    ReadProcessMemory
    ResetEvent
    SetConsoleActiveScreenBuffer
    SetConsoleCP
    SetConsoleCtrlHandler
    SetConsoleCursorPosition
    SetConsoleMode
    SetConsoleOutputCP
    SetConsoleScreenBufferSize
    SetConsoleTextAttribute
[A] SetConsoleTitleA (SetConsoleTitleW)
    SetEndOfFile
[A] SetEnvironmentVariableA (SetEnvironmentVariableW)
    SetEnvironmentVariableW
    SetErrorMode
[A] SetFileAttributesA (SetFileAttributesW)
    SetFilePointer
    SetFileTime
    SetHandleInformation
    SetLastError
    SetSystemTime
    SetUnhandledExceptionFilter
    Sleep
    SleepEx
    TerminateProcess
    TlsGetValue
    UnmapViewOfFile
    VirtualProtect
    VirtualQuery
    WaitForMultipleObjects
    WaitForSingleObject
    WideCharToMultiByte
    WriteConsoleW

msvcrt.dll:
    __C_specific_handler
    ___mb_cur_max_func
[A] __argv (__wargv)
[A] __getmainargs (__wgetmainargs)
[A] __initenv (__winitenv)
    __iob_func
    __set_app_type
    __setusermatherr
[A] _access (_waccess)
    _amsg_exit
    _cexit
    _commode
[A] _ctime64 (_wctime64)
[A] _environ (_wenviron)
    _errno
    _exit
[A] _fdopen (_wfdopen)
    _filbuf
    _fmode
    _fstat64
[A] _fullpath (_wfullpath)
    _get_osfhandle
[A] _getch (_getwch)
    _gmtime64
    _initterm
    _isatty
    _localtime64
    _lseeki64
    _mktime64
    _onexit
[A] _open (_wopen)
    _open_osfhandle
    _pipe
[A] _putenv (_wputenv)
    _setjmp
    _setmode
    _stricmp
    _strnicmp
    _telli64
    _time64
    _timezone
    _tzset
[A] _vscprintf (_vscwprintf)
[A] _vsnprintf (_vsnwprintf)
    _wspawnve
    abort
    atof
    atoi
    bsearch
    calloc
    clearerr
    clock
    exit
    fclose
    feof
    ferror
    fflush
[A] fgets (fgetws)
[A] fopen (_wfopen)
[A] fprintf (fwprintf)
[A] fputc (fputwc)
[A] fputs (fputws)
    fread
    free
    fseek
    fwrite
[A] getenv (_wgetenv)
[A] isalnum (iswalnum)
[A] isalpha (iswalpha)
[A] iscntrl (iswcntrl)
[A] isgraph (iswgraph)
[A] islower (iswlower)
[A] isprint (iswprint)
[A] ispunct (iswpunct)
[A] isspace (iswspace)
[A] isupper (iswupper)
[A] isxdigit (iswxdigit)
    localeconv
    longjmp
    malloc
    mbstowcs
    memchr
    memcmp
    memmove
[A] putc (putwc)
[A] putchar (putwchar)
    qsort
    rand
    realloc
[A] rename (_wrename)
    setbuf
[A] setlocale (_wsetlocale)
    signal
[A] sprintf (swprintf)
    srand
[A] sscanf (swscanf)
[A] strcat (wcscat)
[A] strchr (wcschr)
[A] strcmp (wcscmp)
[A] strcpy (wcscpy)
[A] strcspn (wcscspn)
    strerror
[A] strftime (wcsftime)
[A] strlen (wcslen)
[A] strncmp (wcsncmp)
[A] strncpy (wcsncpy)
[A] strpbrk (wcspbrk)
[A] strrchr (wcsrchr)
[A] strspn (wcsspn)
[A] strstr (wcsstr)
[A] strtok (wcstok)
[A] strtol (wcstol)
[A] strtoul (wcstoul)
[A] tolower (towlower)
[A] toupper (towupper)
[A] ungetc (ungetwc)
[A] vfprintf (vfwprintf)
    wcschr
    wcscpy
    wcslen
    wcsncmp
    wcstombs
    _strtoui64
    _strtoi64
    _write
    _wcsnicmp
[A] _unlink (_wunlink)
    _umask
    _tzset
    _strdup
[A] _spawnve (_wspawnve)
[A] _rmdir (_wrmdir)
    _read
[A] _open (_wopen)
[A] _mktemp (_wmktemp)
[A] _mkdir (_wmkdir)
    _getpid
[A] _getcwd (_wgetcwd)
[A] _getche (_getwche)
    _fileno
[A] _fdopen (_wfdopen)
    _dup2
    _dup
[A] _creat (_wcreat)
    _close
[A] _chmod (_wchmod)
[A] _chdir (_wchdir)
    _stricmp

SHELL32.dll:
    CommandLineToArgvW

USER32.dll:
[A] DispatchMessageA (DispatchMessageW)
    MsgWaitForMultipleObjects
[A] PeekMessageA (PeekMessageW)
    TranslateMessage

WS2_32.dll:
    WSACleanup
    WSAEnumNetworkEvents
    WSAEventSelect
    WSAGetLastError
    WSASetLastError
[A] WSASocketA (WSASocketW)
    WSAStartup
    __WSAFDIsSet
    accept
    bind
    closesocket
    connect
    freeaddrinfo
    getaddrinfo
    gethostbyaddr
    gethostname
    getnameinfo
    getpeername
    getservbyname
    inet_addr
    inet_ntoa
    listen
    recv
    select
    setsockopt
    shutdown
```
</details>

I'm guessing a good bunch of the ansi APIs don't need special handling, e.g. many IO APIs, like `fprintf`, already work fine with UTF8 too (i.e. no need to add a UTF8 wrapper) , and similarly for string APIs like `strlen` (which has an equivalent `wcslen`), or we probably don't need `iswlower`, etc.

Once we have native UTF8 working, we can add a white list of allowed ansi symbols, and then use the script occasionally to check if we need to add more UTF8 wrappers etc.

So this is largely an FYI, but I'd appreciate if you could glance over it and see if anything stands out as wrong.

For reference, here's the current script:  [w32sym.zip](https://github.com/rmyorston/busybox-w32/files/12386408/w32sym.zip)

**Comment by avih** [comment]
> I'm guessing a good bunch of the ansi APIs don't need special handling...

Yeah, I think I overshoot a bit with the matching wide symbols.

These are correct matching symbols (e.g. `strlen`, `wcslen`) but they're only different because they operate on different type.

Compared, for instance, to `[_w]fopen` or `CreateFile{A,W}`, which have different reach - the ANSI ones are more limited compared to the wide ones, and these are the ones we mainly care about.

<details>
<summary>So here's the revised shorter list of ansi symbols + wide matches</summary>

```
$ w32sym.sh -a ./busybox.exe | grep '\['
[A] GetUserNameA (GetUserNameW)
[A] CreateEventA (CreateEventW)
[A] CreateFileA (CreateFileW)
[A] CreateFileMappingA (CreateFileMappingW)
[A] CreateNamedPipeA (CreateNamedPipeW)
[A] CreateProcessA (CreateProcessW)
[A] FillConsoleOutputCharacterA (FillConsoleOutputCharacterW)
[A] FindFirstFileA (FindFirstFileW)
[A] FindNextFileA (FindNextFileW)
[A] GetCompressedFileSizeA (GetCompressedFileSizeW)
[A] GetDiskFreeSpaceExA (GetDiskFreeSpaceExW)
[A] GetDriveTypeA (GetDriveTypeW)
[A] GetFileAttributesA (GetFileAttributesW)
[A] GetFileAttributesExA (GetFileAttributesExW)
[A] GetFullPathNameA (GetFullPathNameW)
[A] GetModuleFileNameA (GetModuleFileNameW)
[A] GetModuleHandleA (GetModuleHandleW)
[A] GetSystemDirectoryA (GetSystemDirectoryW)
[A] GetVersionExA (GetVersionExW)
[A] GetVolumeInformationA (GetVolumeInformationW)
[A] LoadLibraryA (LoadLibraryW)
[A] LoadLibraryExA (LoadLibraryExW)
[A] MoveFileExA (MoveFileExW)
[A] Process32First (Process32FirstW)
[A] Process32Next (Process32NextW)
[A] ReadConsoleInputA (ReadConsoleInputW)
[A] SetConsoleTitleA (SetConsoleTitleW)
[A] SetEnvironmentVariableA (SetEnvironmentVariableW)
[A] SetFileAttributesA (SetFileAttributesW)
[A] __argv (__wargv)
[A] __getmainargs (__wgetmainargs)
[A] __initenv (__winitenv)
[A] _access (_waccess)
[A] _ctime64 (_wctime64)
[A] _environ (_wenviron)
[A] _fdopen (_wfdopen)
[A] _fullpath (_wfullpath)
[A] _open (_wopen)
[A] _putenv (_wputenv)
[A] fopen (_wfopen)
[A] getenv (_wgetenv)
[A] rename (_wrename)
[A] setlocale (_wsetlocale)
[A] _unlink (_wunlink)
[A] _spawnve (_wspawnve)
[A] _rmdir (_wrmdir)
[A] _open (_wopen)
[A] _mktemp (_wmktemp)
[A] _mkdir (_wmkdir)
[A] _getcwd (_wgetcwd)
[A] _fdopen (_wfdopen)
[A] _creat (_wcreat)
[A] _chmod (_wchmod)
[A] _chdir (_wchdir)
[A] DispatchMessageA (DispatchMessageW)
[A] PeekMessageA (PeekMessageW)
[A] WSASocketA (WSASocketW)
```
</details>

And here's the revised script. By default it now produces the short list (limited ANSI APIs), but it can also produce the long list with the general ansi APIs: [w32sym.v2.zip](https://github.com/rmyorston/busybox-w32/files/12394278/w32sym.v2.zip)

But I'm guessing it can't detect symbols which are loaded dynamically? Not sure...

**Comment by rmyorston** [comment]
>Would you be against moving libbb.h it to the end?

Not if it's strictly necessary.

>But I'm guessing it can't detect symbols which are loaded dynamically? Not sure...

I guess not. They'll appear as strings in the binary not as symbols. Easy enough to find in the source, though.

**Comment by avih** [comment]
> * Have the shell spawn an application that spawns some other application. In the original issue the first application was GNU make and I used busybox-w32 `time` as the second.

Hmm.. I can't reproduce this.

I've built busybox.exe with UCRT on windows, using [winlibs](https://github.com/brechtsanders/winlibs_mingw), i386, adding to the path a dir with busybox and the output of `busybox.exe --install .` minus `ar`, plus `xxd`, `pkg-config` and `make` from the w64devkit bin dir (it probably doesn't need `make` because winlibs has it, but it does seem to need a non-busybox xxd).

I got many warnings where LL_FMT and OFF_FMT are used - these seem to be ignored (it thinks `"...%"LL_FMT"d"` is `%d`, and I'm not sure why, but I know that it definitely goes through the `#define LL_FMT "I64"` line), but it did build, and it does depend/linked with `ucrtbase.dll`.

I then used a (wide) spawn variant which replaces NULL env with environ, so that spawn is never called with a NULL env, and does get called with environ env.

I couldn't quite figure out how to setup the make/time example, so I tried these, and all seem to work:
- `./busybox sh -c './busybox sh -c busybox'`
- `./busybox sh -c 'sh -c "sh -c busybox"'`
- `./busybox sh -c 'sh -c "./busybox printf xxx | ./busybox grep xxx"'`

It would really help to have some one-liner test case to reproduce it...

Anyway, for now, I'll keep the current UCRT code which uses NULL env, and I'll handle exporting the UTF8 env to the system env, because that would also typically be much quicker to convert only he unicode values instead of converting the whole UTF8 environ to wide. It's not a big function so that should be fine.

But I would still like to be able to reproduce the UCRT crash when passing environ env, and then confirm that the NULL env fixes it.

**Comment by avih** [comment]
> I tried these, and all seem to work:
> 
> `./busybox sh -c './busybox sh -c busybox'`
> `./busybox sh -c 'sh -c "sh -c busybox"'`
> `./busybox sh -c 'sh -c "./busybox printf xxx | ./busybox grep xxx"'`

I tried also with the latest winlibs x86-64, plain non-any-utf8 build, with only this diff to the source:
```diff
diff --git a/shell/ash.c b/shell/ash.c
index 5a5c947e8..ec11e8f56 100644
--- a/shell/ash.c
+++ b/shell/ash.c
@@ -9126,7 +9126,7 @@ tryexec(IF_FEATURE_SH_STANDALONE(int applet_no,) const char *cmd, char **argv, c
 # else
 		if (APPLET_IS_NOEXEC(applet_no)) {
 # endif
-#if ENABLE_PLATFORM_MINGW32 && !defined(_UCRT)
+#if 1 || ENABLE_PLATFORM_MINGW32 && !defined(_UCRT)
 			/* If building for UCRT move this up into shellexec() to
 			 * work around a bug. */
 			clearenv();
@@ -9203,7 +9203,7 @@ static void shellexec(char *prog, char **argv, const char *path, int idx)
 	int applet_no = -1; /* used only by FEATURE_SH_STANDALONE */
 
 	envp = listvars(VEXPORT, VUNSET, /*strlist:*/ NULL, /*end:*/ NULL);
-#if ENABLE_PLATFORM_MINGW32 && defined(_UCRT)
+#if 0 && ENABLE_PLATFORM_MINGW32 && defined(_UCRT)
 	/* Avoid UCRT bug by updating parent's environment and passing a
 	 * NULL environment pointer to execve(). */
 	clearenv();
```

And tried also this case:
```sh
./busybox sh -c 'X=x ./busybox sh -c "Z=z ./busybox"'
```
Which also works fine as far as I can tell..

**Comment by rmyorston** [comment]
I'm also unable to reproduce the problem. Maybe Microsoft have fixed UCRT.

**Comment by avih** [comment]
> 

OK, I'll leave it up to you to decide what to do with the UCRT workaround.

Meanwhile, I'll just support the standard behavior with (utf8) spawn: allow and "forward" both NULL and non-NULL env, and if it's NULL then ensure the system wide env is up to date (instead of the smaller and slower hack of replacing NULL with environ and then converting all of it to wide).

As for the UCRT I64 warnings, I have a patch and will send a PR soon.

I'm Just confirming first it compiles without warnings with/without UCRT for i686/x86-64.

**Comment by rmyorston** [comment]
OK, this rabbit hole was a bit deeper than expected.

It bothered me that I was unable to reproduce the problem with UCRT. The original issue (#234) was related to an interaction between GNU make and busybox-w32 when they were both compiled with UCRT.

I was able to reproduce the problem using GNU make from back then but wanted to do it solely with current busybox-w32 compiled with the current MinGW-w64 UCRT toolchain.

The issue in GNU make is that it passes a non-default environment to `CreateProcess()`. In busybox-w32 `CreateProcess()` is used in the implementation of `popen(3)`, so I hacked that to have a non-default environment:
```diff
diff --git a/win32/popen.c b/win32/popen.c
index 2208aa6bb..de2f2b1fb 100644
--- a/win32/popen.c
+++ b/win32/popen.c
@@ -190,6 +190,7 @@ static int mingw_popen_internal(pipe_data *p, const char *cmd,
    int success;
    int fd = -1;
    int ip, ic, flags;
+   char env[] = "HELLO=1\0";

    if ( cmd == NULL || *cmd == '\0' || mode == NULL ) {
        return -1;
@@ -251,7 +252,7 @@ static int mingw_popen_internal(pipe_data *p, const char *cmd,
                NULL,              /* primary thread security attributes */
                TRUE,              /* handles are inherited */
                0,                 /* creation flags */
-               NULL,              /* use parent's environment */
+               env,
                NULL,              /* use parent's current directory */
                &siStartInfo,      /* STARTUPINFO pointer */
                &p->piProcInfo);   /* receives PROCESS_INFORMATION */
```
Then I modified the shell to force it to spawn all applets and reverted the UCRT workaround in the shell:
```diff
diff --git a/shell/ash.c b/shell/ash.c
index 2ea87a049..0f1862424 100644
--- a/shell/ash.c
+++ b/shell/ash.c
@@ -9103,7 +9103,7 @@ tryexec(IF_FEATURE_SH_STANDALONE(int applet_no,) const char *cmd, char **argv, c
 {
 #if ENABLE_FEATURE_SH_STANDALONE
    if (applet_no >= 0) {
-# if ENABLE_PLATFORM_MINGW32
+# if ENABLE_PLATFORM_MINGW32 && 0
        /* Treat all applets as NOEXEC, including the shell itself if
         * this is a FS_SHELLEXEC shell. */
        struct forkshell *fs = (struct forkshell *)sticky_mem_start;
@@ -9124,7 +9124,7 @@ tryexec(IF_FEATURE_SH_STANDALONE(int applet_no,) const char *cmd, char **argv, c
 # else
        if (APPLET_IS_NOEXEC(applet_no)) {
 # endif
-#if ENABLE_PLATFORM_MINGW32 && !defined(_UCRT)
+#if ENABLE_PLATFORM_MINGW32 && !defined(_WHATEVER)
            /* If building for UCRT move this up into shellexec() to
             * work around a bug. */
            clearenv();
@@ -9201,7 +9201,7 @@ static void shellexec(char *prog, char **argv, const char *path, int idx)
    int applet_no = -1; /* used only by FEATURE_SH_STANDALONE */

    envp = listvars(VEXPORT, VUNSET, /*strlist:*/ NULL, /*end:*/ NULL);
-#if ENABLE_PLATFORM_MINGW32 && defined(_UCRT)
+#if ENABLE_PLATFORM_MINGW32 && defined(_WHATEVER)
    /* Avoid UCRT bug by updating parent's environment and passing a
     * NULL environment pointer to execve(). */
    clearenv();
```
To make the call to `popen(3)` I used this Makefile:
```
sum != sum Makefile

target:
	@echo sum is $(sum)
```
The shell macro assignment (`!=`) uses `popen(3)`.  Using a UCRT-compiled version of busybox-w32 `make` from current git master gives:
```
~ $ make
sum is 22355 1
```
The UCRT version without the workaround gives:
```
~ $ make
sum is
```
So, the workaround is still necessary. Without it GNU make still triggers the problem previously reported.

**Comment by avih** [comment]
Here are some more thoughts (i'm not abandoning the native utf8, but still).

Mingw-w64 recently changed their default to ucrt (at git master, probably for v12, if there wouldn't be too much backlash), and while w64devkit doesn't currently have plans to switch to ucrt, I'm guessing the defaults would eventually trickle down to most mingw setups.

ucrt applications can run on XP. It does require installing the ucrt runtime (the is/was an official runtime for XP), but then it works, even when compiling with llvm 17 (with some XP compiler and linker options - if the mingw setup defaults to win 7 or later).

Using a utf8 locale does not actually require the utf8 manifest. It can also be set using `setlocale(".UTF8")`. Similar to the manifest, this would only succeed on win10 1803+ (the manifest is 1903+ - not the same version), but unlike the manifest, it also runs on XP, and if setlocale failed then no problem - it remains ANSI.

Another difference with runtime `setlocale` is that `argv` and `environ` are already ANSI - unlike with the manifest, so they need to become utf8. Also, apparently, `CP_ACP` and `GetACP()` are as if setlocale was never called (but of course, the current locale - which is affected by setlocale - can be determined at runtime).

And, also unlike the manifest, this only works with ucrt. With msvcrt on win 10/11 (and older versions) the setlocale call fails, which is why I mentioned the mingw default change and that ucrt can run on XP as well.

Here's an example of using the runtime setlocale approach: https://github.com/gwsw/less/commit/ad440b6cea6a8ac2c5c26011fe4fe710791b459f

It's relatively little code. More than the manifest (which basically requires no code changes), but not too much. I think it can trivially work in busybox-w32 as well.

A manifest can still be added, which would bypass the initial conversion of argv and environ, but I'm guessing that the goal of this is to support XP up to latest with the same binary, and the manifest breaks XP.

And, we might need to reconsider our position about the same binary being unicode on one system and ansi on another.

Thoughts?

**Comment by avih** [comment]
> Using a utf8 locale does not actually require the utf8 manifest. It can also be set using `setlocale(".UTF8")`. Similar to the manifest

So, I have a local patch to use this method as FEATURE_UTF8_UCRT, but it seems that it doesn't work, or I couldn't make it work, as I expected.

TL;DR: `fopen` seems to work and does open UTF-8 names correctly, and the [docs](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/setlocale-wsetlocale?view=msvc-170) mention that `_mkdir` and `_getcwd` do become UTF-8, and (with this busybox build) the `spawn`* family does invoke apps with correct unicode args (i.e. seemingly argv is converted correctly to UTF8 by new code in this patch, and `spawn` uses this UTF8 without further conversion to wide chars), but as far as I can tell `Find{First,Next}FileA` seem to return question marks instead of non-ASCII codepoints.

I've reproduced the `fopen` and `FindFirstFileA` cases in a test program, where with `setlocale` only the former becomes UTF8, while with the UTF8 manifest both seem to become UTF8.

See also here https://github.com/gwsw/less/pull/438

I don't know whether that's a bug of the UTF8 locale or a designed behavior - I don't think there's docs for this locale other than the linked page.

**Comment by avih** [comment]
> and I don't quite get where the line goes between the manifest and the setlocale methods (other than argv, environ, etc).

Or maybe I do. I'm guessing the `setlocale` method only applies to msvcrt/ucrt API, like `fopen`, but indeed not like `FindFirstFileA`.

Basically, roughly the libc API, which I guess kind of makes sense, but is not very helpful if one needs total UTF8 API, including the ANSI win32 APIs...

**Comment by ale5000-git** [comment]
@rmyorston
@avih
Wouldn't it be possible to allow to user to enable/disable the interactive UTF-8 editing at runtime using a special BB_* var?

**Comment by avih** [comment]
Please stop it. The answer is no. Just replace the binary instead.

**Comment by idigdoug** [comment]
I just finished the first draft of support for native encoding support. It seems to work ok -- it passes and fails the exact same set of tests as the manifest-utf-8 build. It works as follows:

- At startup, select a code page based on environment variable. If no environment variable is set, or if the specified code page is not supported, use utf-8.
- Whenever BB wants to interact with a Windows API, it converts all strings to UTF-16 using the specified code page, calls the W version of the Windows API, then converts any returned strings back to CHAR using the specified code page.
- Same with CRT APIs -- for APIs that interact with filenames, always use the WIDE API.

Once I've tested this a bit, I'll make a PR and we can discuss details.

**Comment by avih** [comment]
> * it converts all strings to UTF-16 using the specified code page, calls the W version of the Windows API, then converts any returned strings back to CHAR using the specified code page.

- How do you enumerate all the A APIs which are currently used by busybox-w32, to know that you covered all of them?
- What happens with functions which modify a `char*` value, like `mingw_mktemp` and `mkstemp` (both at `mingw.c`)?
- How is `argv` handled?
- Does it ensure that env vars with unicode names and/or values are handled correctly?

**Comment by idigdoug** [comment]
1. On windows, you run `link /dump /imports busybox.exe > bb.imports` and you see all of the APIs that are imported from DLLs. (Doesn't catch LoadLibrary, but gets everything else.) This allows audits. At present, my binary uses no Windows -A APIs. CRT APIs are harder to audit, but I've checked all of them.
2. You convert to W on the way in, then convert back to A on the way back out.
3. There are alternatives, and I'm not stuck on any specific implementation. Easy to change. My current implementation uses the wmain entry point so it gets `wchar_t** wargv`. Then I convert that back to the requested encoding (default UTF-8) to make `char** argv` and use that for the rest of the program. (Also set `__argv = convertedArgv` so that getopt is happy.)
4. Yes. The OS copy of the environment block is kept up to date with SetEnvironmentVariableW. I maintain my own `environ` block, and `getenv` returns those values. I do have to use the CRT's native environment block to set the TZ variable for use in a few date/time routines.

**Comment by avih** [comment]
> 1. On windows, you run `link /dump /imports busybox.exe > bb.imports` and you see all of the APIs that are imported from DLLs

That requires msvc `link`, and while there are free versions of Visual Studio, preferably it should not depend on msvc, and instead only use open source tools, like `objdump` in binutils.

**Comment by idigdoug** [comment]
You can use `objdump -p busybox.exe` to dump a lot of information, including imports. Search the output for "The Import Tables" and "DLL Name:".

First draft of my change is here: https://github.com/idigdoug/busybox-w32/tree/wide_apis

Passes the same number of tests as mingw64u_defconfig without my changes: "64 failure(s) detected".

Still lots of manual testing to do.

Still lots of code to self-review.

Still lots of decisions to be made (feedback welcome):

- At present, encoding defaults to UTF-8, can be overridden at process start via BB_CODEPAGE environment variable.
- Unicode functions (unicode.c) are always enabled and always work based on the selected encoding (cannot disable via environment or locale).

Binary sizes (mingw64u_defconfig):

- Old: 708,096
- New: 719,872

Status:

- I have not yet put any effort into enabling the non-Unicode config. Only mingw64u_defconfig will work. (Specifically, things will probably go horribly wrong if `CONFIG_UNICODE_SUPPORT` is not enabled.)
- Several of the UTF8-related config options are now meaningless. I have not yet removed them from Kconfig.
  - CONFIG_FEATURE_UTF8_MANIFEST is probably useless. I've turned it off.
  - CONFIG_FEATURE_FAIL_IF_UTF8_MANIFEST_UNSUPPORTED is useless.
  - CONFIG_FEATURE_UTF8_INPUT and CONFIG_FEATURE_UTF8_OUTPUT are dead.
  - CONFIG_FEATURE_CHECK_UNICODE_IN_ENV is no longer used (mbstowcs/wcstombs and related are hard-coded to "on" and always function according to the process's encoding).

**Comment by rmyorston** [comment]
@idigdoug Thanks for your efforts!

I'll have a look, but I can't promise to say anything intelligent about them.

I'm sure @avih will be along shortly to do that.

**Comment by avih** [comment]
> I'm sure [@avih](https://github.com/avih) will be along shortly to do that.

I'm not so sure.

It feels to me like the changes are way more intrusive than needed, and the commit messages literally don't say anything at all.

There are barely comments at the code, nothing is explained anywhere, no choice is discussed anywhere, and the reader is expected to figure out everything (behavior, reasons, and correctness) by staring at more than 2000 changes all over the place.

Not really my cup of tea, to be honest.

@idigdoug did you write it all by yourself or did you use some LLMs? if LLM were involved, which ones were used, and how much is written by LLMs and how much by you (roughly, percentage wise)?

**Comment by avih** [comment]
Also, correct me if I'm wrong, but it looks to me like all or the vast majority of changes are unconditional. I.e. there's no way to build non-unicode busybox-w32 anymore.

Also, it's unclear to me what the `mingw64u_defconfig` now does (or the other configs, eventhough they're untouched), and why the UTF8_INPUT and UTF8_OUTPUT are disabled.

EDIT: it looks to me like UTF8_INPUT and UTF8_OUTPUT were extended to general Unicode handling and are now unconditional, at which case these two configs should have been removed rather than disabled, at least if it's acceptable that Unicode becomes unconditional - which looks to me like what the code does.

As I said, nothing is explained, and no choice/decision is described or argued.

**Comment by rmyorston** [comment]
One thing that struck me is code like:
```c
    {
        int saved_errno = errno;
        mingw_wcs_free(&wr);
        errno = saved_errno;
    }
```
I hadn't realised `free(3)` in the Microsoft C runtime can change `errno`. Upstream BusyBox, of course, assumes the behaviour of `free(3)` in glibc, where it's documented to preserve `errno`. (I couldn't immediately find documentation on this point for musl, but the code certainly takes steps to preserve `errno`.)  I see I've assumed the same in code I've written for busybox-w32.

How about a `mingw_free()` wrapper which preserves `errno`? This would align with the assumptions of the current code and avoid the need for code like the above.

**Comment by idigdoug** [comment]
As I said before, is a first draft. I shared it now because I need feedback at this point for decisions. It is not yet ready for PR.

As I said before, I haven't yet tried to make this conditional or remove the no-longer-relevant config options. I am primarily interested in feedback for how high of a priority it is to keep the existing "use the CRT and OS charset translation" behavior as an option since it would be quite a bit of work going forward to maintain both paths.

I think all of the changes are necessary if we're aiming for correctness. If you have a better way to do it that is less-invasive and still correct then go ahead. I'm just trying to help. I'm a little bit surprised at the hostility.

I typically do my write-up in the pull request, not in individual commits.

No, free doesn't change errno. This was a mistake on my part.

Yes, I used an LLM to help with my work: Microsoft Copilot which is a front-end to other LLMs. Claude was used to help with the first-draft of the code, and GPT was used as rubber-duck. I have carefully gone through the code outside of win32 and I think I'm happy with it. I have not yet carefully gone through the changes inside of win32 and there may be additional changes there.

**Comment by avih** [comment]
> I'm just trying to help.

That's great.

> I'm a little bit surprised at the hostility.

Which parts do you consider hostile?

> is a first draft... It is not yet ready for PR.

That's obvious to me, or at least I'd consider it as such, but since this is your first PR to this project, and it's a very big PR, we don't know how capable you are (hopefully very, but we don't know that), and I've seen enough LLM-powered PRs in the past which largely ended up as a waste of time of the maintainers, I pointed out the sore points I see from a very high overview.

EDIT: it's not actually a PR yet, so I can only comment on the existing code and commits in your branch, and obviously I can't know how your PR would look like.

> I am primarily interested in feedback for how high of a priority it is to keep the existing "use the CRT and OS charset translation" behavior as an option

I might have missed that, but I've not seen such request about this specific subject prior to this comment, and if that's indeed the case, why didn't you ask for it in advance?

Are there other specific aspects you'd like to hear feedback for?

I think it should remain an option, so that we still have the ANSI/UTF8-manifest variants which are battle-proven by now, and also be able to isolate whether some issues are inherently busybox issues, or specifically related to the unicode support, though ultimately it's @rmyorston decision to make.

> since it would be quite a bit of work going forward to maintain both paths.

That is correct, which is where my first comment comes in:
> It feels to me like the changes are way more intrusive than needed

I feel that correct integration would help with both causes: reduce dramatically the required changes, and also make it much easier to keep it optional/configurable.

Such a big change would never be trivial to integrate, but I think it can and should be much less intrussive to the existing codebase than what the current commits do.

> I typically do my write-up in the pull request, not in individual commits.

Users who clone the repo and browse the git history don't see the PR writeup. They see individual commits, which are composed of commit titles, commit messages, and source changes per commit.

The PR writeup may give some overview of the whole thing (AKA "cover letter"), but ultimately the commits themselves are your only output, and should include everything you think another developer might be interested at or should be aware of. Changes should be described/reasoned/etc at the commit message and summarized at the commit title, while comments regarding the (new) code should be at the source code.

There are enough places which describe how to write good commit messages, and what should be a reasonable scope for a single commit. In general, every commit should do one thing, and should be simple to follow as much as possible, but it's not an exact science.

> Yes, I used an LLM to help with my work

Rather than a reply to my question, this is something I'd have very much preferred to see disclosed voluntarily (or even mandatorily, as many projects require these days, though not yet this one AFAIK).

It has implications in various aspects, like licensing (I've identified my very own code as the basis of several LLM-powered PRs - elsewhere, with infringment and zero attribution), how much is the "author" capable of handling review comments on their own (hopefully very capable, but unfortunately I've seen too many examples to the contrary), how much is the author capable of solving this problem on their own (many times they're not), and how committed is the author to maintain their changes in the future (we can't ever know in advance, but it's more likely with someone who invested meaningful time, solved real problems, and made hard decisions with this codebase).

**Comment by idigdoug** [comment]
In my message from last night, I pre-emptively mentioned several of the issues you brought up:

---

Status:

- I have not yet put any effort into enabling the non-Unicode config. Only mingw64u_defconfig will work. (Specifically, things will probably go horribly wrong if CONFIG_UNICODE_SUPPORT is not enabled.)
- Several of the UTF8-related config options are now meaningless. I have not yet removed them from Kconfig.
  - CONFIG_FEATURE_UTF8_MANIFEST is probably useless. I've turned it off.
  - CONFIG_FEATURE_FAIL_IF_UTF8_MANIFEST_UNSUPPORTED is useless.
  - CONFIG_FEATURE_UTF8_INPUT and CONFIG_FEATURE_UTF8_OUTPUT are dead.
  - CONFIG_FEATURE_CHECK_UNICODE_IN_ENV is no longer used (mbstowcs/wcstombs and related are hard-coded to "on" and always function according to the process's encoding).

---

I didn't specifically state the areas where I wanted feedback, but I definitely want feedback on the areas I mentioned in my second response.

Regarding commit write-ups: I have been doing commits for 25 years. When I'm still on a first draft, I don't spend a lot of time writing up commits. When I am ready to do a PR, I expect to do a squash commit with a nice write-up.

If you have specific suggestions about how this could be less-invasive and still be correct, I'm happy to hear them. I currently think this is the minimum set of changes you can have to be "correctly" handling text encodings.

>> I might have missed that, but I've not seen such request about this specific subject prior to this comment, and if that's indeed the case, why didn't you ask for it in advance?

Several months ago I mentioned that I might want to work on this project, but I said I would like to plan a bit with maintainers before starting work. You told me to either shut up or get to work. So I did what you asked and got to work instead of planning ahead with maintainers.

https://github.com/rmyorston/busybox-w32/issues/386#issuecomment-2725351706

**Comment by avih** [comment]
> In my message from last night, I pre-emptively mentioned several of the issues you brought up:

Right. Thanks. I somehow overlooked some of it.

> CONFIG_FEATURE_CHECK_UNICODE_IN_ENV is no longer used

I think this one has some use cases, though I don't recall the exact specifics. IIRC it's an upstream busybox feature which allows changing LC_ALL and/or LC_LANG to affect the behavior of some things, which the UTF8-manifest build keeps and allows similar control on Windows.

> I didn't specifically state the areas where I wanted feedback, but I definitely want feedback on the areas I mentioned in my second response.

Is there anything left without feedback from me?

> Regarding commit write-ups: I have been doing commits for 25 years.

That's great, but we couldn't have known that. I was judging what I saw - which is the current set of commits at the branch `wide_apis`.

> When I am ready to do a PR, I expect to do a squash commit with a nice write-up.

With this scale of code changes, I'd expect a good PR to include quite a few more than a single squashed commit. See my previous comment about how I (and others) think PR and commits should look like.

> You told me to either shut up or get to work.

It's actually a rather detailed and hopefully useful [comment](https://github.com/rmyorston/busybox-w32/issues/386#issuecomment-2725351706), with a bottom line of "I'm done with it for now, if you want it, now is your turn to have a go".

It does ask an important question which I've not seen answered yet: how is it better than the UTF8-manifest solution (beyond win XP/7/8)? and do you think this difference, if it exists, is worth so much code changes?

**Comment by idigdoug** [comment]
As mentioned in the prior thread:

- Ability to work with other encodings using the same busybox binary rather than needing different binaries for UTF-8 and ACP encodings. A lot of text is not UTF-8. Being able to do things like use `busybox xargs` to launch processes based on non-UTF-8 input seems useful.
- Able to work on older versions of Windows. This is not important to me, but seems to be important to some other people.
- I know there have been several problems with the CRT support for .UTF8 locales, e.g. spawn. This change avoids those issues.

In addition:

- wchar_t conversion is needed to resolve long-path issues (https://github.com/rmyorston/busybox-w32/issues/562) so if we want long-path support we need some infrastructure for handling wchar_t encodings for path names.
- Would fix https://github.com/rmyorston/busybox-w32/issues/399
- Might fix others.

I have no issue with adding back config features that are useful. First draft just hard-coded them for proof of concept.

CONFIG_FEATURE_CHECK_UNICODE_IN_ENV is from upstream, but I'm still a bit fuzzy on when it is useful. So for first draft, I disabled it. I'm definitely open to enabling it and any other features that need to be kept.

**Comment by avih** [comment]
> * Ability to work with other encodings using the same busybox binary rather than needing different binaries for UTF-8 and ACP encodings. A lot of text is not UTF-8. Being able to do things like use `busybox xargs` to launch processes based on non-UTF-8 input seems useful.

Not sure I understand this one. This statement suggest that currently such problems are solvable, but using two binaries (ACP and UTF8-manifest), right?

Do you have a specific real-world use case which is not possible to do with the UTF8-manifest binary but is possible with the ACP binaries?

Also, xargs only creates an arguments array. If the arguments values are correct - which I think it should be with the UTF8-manifest solution, then it doesn't matter whether the launched application is purely ACP. It will still be launched with the correct `wchar_t` values, which would translate correctly - as much as possible - to ACP which such utility might expect. And same with input arguments to busybox with the UTF8-manifest.

> * Able to work on older versions of Windows. This is not important to me

This is the main reason _I_ think of, and while it's not negligible, its value also decreases rapidly as none of these systems are still supported today, and even less so tomorrow.

> * I know there have been several problems with the CRT support for .UTF8 locales, e.g. spawn. This change avoids those issues.

Please share the ones you know of which affects the UTF8-manifest busybox-w32 binaries. If possible, we'd like to address them, or at least be aware of them. Thanks.

> * wchar_t conversion is needed to resolve long-path issues

To solve this problem, a MUCH smaller solution would likely suffice, preferably independent of the "full unicode" support. _That_ could be a useful first PR, which would likely be much smaller in scale than "full uniocode". Such PR does exist - #564, feel free to help improve it.

> * Would fix [32-bit unicode support #399](https://github.com/rmyorston/busybox-w32/issues/399)

Already fixed with the `mingw32w_defconfig` config, and regardless, it was merely "release optimization" to reduce the number of builds. Nothing limits the UTF8-manifest to 64 bit builds, and it's trivial to take an existing UTF8-manifest config and use it to build a 32bit binary - which is what the "32w" config does.

> * Might fix others.

Maybe, but the question is which ones.

Preferably we need to know that to be able to assess the usefulness of such change, right?

**Comment by ale5000-git** [comment]
This is my suggestion:

- Add the possibility to switch UTF-8 on/off at runtime similar to linux using the LANG env var.
- Set LANG to `en_US.UTF-8` by default if not already set, otherwise use the value already set and then check the presence of `.UTF-8` to enable it.

**Comment by idigdoug** [comment]
Yes, if you only want to work using ACP and UTF-8, then you can solve the problem with two binaries. If you want to work with anything else then you have no solution. I don't like having to switch binaries just to switch code pages, and you're still stuck with just ACP and UTF-8.

xargs creates an argument array by parsing some data (in some encoding) and converting that into a command line. On Unix, that's a binary conversion because command line is char. On Windows, that's a text conversion because command line (and environment) is natively wchar_t. If the input data is utf-8 then you can do this with a utf-8 build of busybox xargs. If the input data is not utf-8, you would have to transcode it to utf-8 before passing it to busybox xargs.

Regarding problems with CRT, I'm having trouble finding the details. I remember reading about the spawn issue https://github.com/rmyorston/busybox-w32/issues/352#issuecomment-1678582758 and finding an analysis that determined this was a bug in the ANSI version of spawn but was not an issue in _wspawnve. However, I cannot find a source for that right now, so I might be misremembering.

Regarding ale5000-git's suggestion, unfortunately I don't know of any way to set the code page of Windows APIs at runtime. They are set either in a global system setting or by the application binary's manifest.

**Comment by avih** [comment]
> Yes, if you only want to work using ACP and UTF-8 ...

> xargs creates an argument array ...

I still don't understand this problem, and I don't want to guess. As requested, please share a real world example which works with the ACP binary but not with the UTF8-manifest binary.

I'm guessing I might be able to synthesize such an example, but as it stands, I don't think it would represent real-world scenarios.

> Regarding problems with CRT, I'm having trouble finding the details. I remember reading about the spawn issue [#352 (comment)](https://github.com/rmyorston/busybox-w32/issues/352#issuecomment-1678582758) and finding an analysis that determined this was a bug in the ANSI version of spawn but was not an issue in _wspawnve. However, I cannot find a source for that right now, so I might be misremembering.

I'm not seeing at this comment anything which might imply that "native unicode" would be better than UTF8-manifest. There was some comment about UCRT which (at the time) had some issue with handling NULL `environ`, and _IIRC_ it was addressed later.

Regardless, if there was a specific issue with the ACP `char*` spawn vs `wchar_t*` spawn, then it's highly likely to be addressable with a spawn-specific fix, which would likely be few lines of code at most.

Again, please share any issues you know of with the UTF8-manifest binaries. We'd like to know about such issues, and preferably try to address them.

Otherwise, I consider such statements FUD. No offense (I mean it - I don't think you try to add FUD, but without facts...).

**Comment by avih** [comment]
> * Ability to work with other encodings using the same busybox binary rather than needing different binaries for UTF-8 and ACP encodings. A lot of text is not UTF-8. Being able to do things like use `busybox xargs` to launch processes based on non-UTF-8 input seems useful.

As noted, I still don't really understand this issue and would still appreciate a real world example of it, however, I'm also guessing that any conversion of values between specific codepages should be covered by busybox-w32 `iconv`, and that would likely also allow solving the issue which that paragpaph describes, albeit likely differently than how the "full unicode" solution would allow solving, but likely with similar amount of effort.

Additionally, if one needs to know the default windows ACP, it could be [achieved](https://stackoverflow.com/a/56922538/3339341) using:
```sh
acp=$(REG QUERY 'HKLM\SYSTEM\CurrentControlSet\Control\Nls\CodePage' -v ACP)
acp=${acp##* }
```
and then used further, as noted, with `iconv`.
